File: 1300-critical-connections-in-a-network/1300-critical-connections-in-a-network.cpp
class Solution {
public:
    vector<vector<int>> adj;
    vector<int> disc;
    vector<vector<int>> critical_conns;
    vector<int> processed;
    int dfs(int i, int depth) {
        if (disc[i] < depth) return disc[i];
        
        disc[i] = depth;
        int min_depth = depth;
        for (int j=0; j<adj[i].size(); j++) {
            if (disc[adj[i][j]] != depth-1 && processed[adj[i][j]] == 0) {
                // we dont want to go back to the parent right away
                int curr_min_depth = dfs(adj[i][j], depth+1);
                if (curr_min_depth > depth) {
                    // bridging edge
                    critical_conns.push_back({i, adj[i][j]});
                } 
                min_depth = min(min_depth, curr_min_depth);
            }
        }

        processed[i] = 1;
        return min_depth;
    }
    vector<vector<int>> criticalConnections(int n, vector<vector<int>>& connections) {
        ios_base::sync_with_stdio(false);cin.tie(nullptr);cout.tie(nullptr);
        adj.resize(n+1, vector<int>(0, {}));
        disc.resize(n+1, INT_MAX);
        processed.resize(n+1, 0);
        for (auto i: connections) {
            adj[i[0]].push_back(i[1]);
            adj[i[1]].push_back(i[0]);
        }
        dfs(0, 1);
        return critical_conns;
    }   
};

File: 1300-critical-connections-in-a-network/1300-critical-connections-in-a-network.py
class Solution:
    def criticalConnections(self, n: int, connections: List[List[int]]) -> List[List[int]]:
        self.adj = [[] for _ in range(n)]
        self.disc = [float('inf')] * n
        self.processed = [0] * n
        self.critical_conns = []

        for u,v in connections:
            self.adj[u].append(v)
            self.adj[v].append(u)
        
        def dfs(i: int, depth: int) -> int:
            if self.disc[i] < depth:
                return self.disc[i]
            
            self.disc[i] = depth
            min_depth = depth
            
            for j in self.adj[i]:
                if self.disc[j] != depth - 1 and self.processed[j] == 0: # not the parent and not been processed:
                    curr_min_depth = dfs(j, depth + 1)
                    if curr_min_depth > depth:
                        # bridging edge has been found
                        self.critical_conns.append([i, j])
                    min_depth = min(min_depth, curr_min_depth)
            
            self.processed[i] = 1
            return min_depth

        dfs(0, 1)
        return self.critical_conns

